includes/validation/class-amp-validation-manager.php

```diff
@@ -182,12 +182,12 @@ public static function init() {
 					'code'    => $should_validate_response->get_error_code(),
 					'message' => $should_validate_response->get_error_message(),
 				],
-				400
+				401
```

Comment:
The `401` code requires a `WWW-Authenticate` to be sent along as well: https://httpstatuses.com/401

I suggest using something like the following:
```
WWW-Authenticate: WordPress realm="wp-login.php", title="Log in via /wp-login.php"
```

The above is a guess based on reading up on the syntax (https://tools.ietf.org/html/rfc7235#section-4.1). Feel free to adapt if you know more about this header tag.

Reply:
The authentication is not performed via logging-in to WordPress however. With the changes in this PR, authentication is always performed via nonce query param. So instead of `WordPress` would it be `nonce`?

Isn't “requires” a bit strong here? WordPress isn't sending back this header. For example:

```
$ curl -i -X PUT -d 'title=hello' 'https://wordpressdev.lndo.site/wp-json/wp/v2/posts/1'
HTTP/2 401
access-control-allow-headers: Authorization, Content-Type
access-control-expose-headers: X-WP-Total, X-WP-TotalPages
allow: GET
content-type: application/json; charset=UTF-8
date: Thu, 12 Dec 2019 07:59:25 GMT
link: <https://wordpressdev.lndo.site/wp-json/>; rel="https://api.w.org/"
server: nginx/1.14.2
x-content-type-options: nosniff
x-powered-by: PHP/7.3.12
x-robots-tag: noindex
content-length: 107

{"code":"rest_cannot_edit","message":"Sorry, you are not allowed to edit this post.","data":{"status":401}}
```

Reply:
Well, WordPress is hardly a good reference.

From the RFC 7235 (Hypertext Transfer Protocol (HTTP/1.1): Authentication):
```
The server generating a 401 response MUST send a WWW-Authenticate header field
(Section 4.1) containing at least one challenge applicable to the target resource.
```
See https://tools.ietf.org/html/rfc7235#section-3.1

Reply:
I don't think anything breaks if we don't send it, but the response would technically be invalid.

Reply:
If I understand the RFCs correctly, `nonce` would be a property of the authentication type `Digest`. And when you send `nonce`, you're supposed to send a new one as a value so the client can retry with a valid nonce:
```
WWW-Authenticate: Digest nonce=12345
```
See https://tools.ietf.org/html/rfc2617#section-3.2.1

Reply:
I don't feel strongly about this, though. Seems like this is all so old and weird and badly documented that no-one gets it right anyway.

Reply:
Just tested with a Google API:
<img width="831" alt="Image 2019-12-12 at 9 20 12 AM" src="https://user-images.githubusercontent.com/83631/70694874-d5884e80-1cc0-11ea-984b-488e61c060f1.png">

Reply:
Google is hardly an authority here! 

😉

Reply:
Well, we might as well go full on board here and actually send the nonce via the `Authorization` request header as well…

Reply:
Humm. Except this would cause problems for sites that are already behind some HTTP authentication and they'd need to provide double authentication.

Reply:
In that case I suppose multiple headers could be sent: https://stackoverflow.com/a/38515091

Reply:
I started going down this road, but I think it's overkill for now.

Stashed patch:

```diff
diff --git a/includes/validation/class-amp-validation-manager.php b/includes/validation/class-amp-validation-manager.php
index b724aae9..feb496c6 100644
--- a/includes/validation/class-amp-validation-manager.php
+++ b/includes/validation/class-amp-validation-manager.php
@@ -19,6 +19,13 @@ class AMP_Validation_Manager {
 	 */
 	const VALIDATE_QUERY_VAR = 'amp_validate';
 
+	/**
+	 * Custom HTTP authorization type that is sent in Authorization request header and WWW-Authenticate response header.
+	 *
+	 * @var string
+	 */
+	const VALIDATE_AUTHORIZATION_TYPE = 'AMP-WP-Validate-Nonce-v1';
+
 	/**
 	 * Query var for containing the number of validation errors on the frontend after redirection when invalid.
 	 *
@@ -1614,7 +1621,7 @@ class AMP_Validation_Manager {
 	 * @return string Nonce.
 	 */
 	public static function get_amp_validate_nonce() {
-		return substr( wp_hash( self::VALIDATE_QUERY_VAR . wp_nonce_tick(), 'nonce' ), -12, 10 );
+		return wp_hash( self::VALIDATE_QUERY_VAR . wp_nonce_tick(), 'nonce' );
 	}
 
 	/**
@@ -1632,11 +1639,44 @@ class AMP_Validation_Manager {
 			return false;
 		}
 
-		$validate_key = wp_unslash( $_GET[ self::VALIDATE_QUERY_VAR ] ); // phpcs:ignore WordPress.Security.NonceVerification.Recommended
-		if ( self::get_amp_validate_nonce() !== $validate_key ) {
+		$validate_nonce = self::get_amp_validate_nonce();
+		$validate_key   = wp_unslash( $_GET[ self::VALIDATE_QUERY_VAR ] ); // phpcs:ignore WordPress.Security.NonceVerification.Recommended
+		if ( $validate_nonce === $validate_key ) {
+			return true;
+		}
+
+		if ( empty( $_SERVER['HTTP_AUTHORIZATION'] ) ) {
+			return new WP_Error(
+				'http_request_failed',
+				esc_html__( 'Missing authorization header.', 'amp' )
+			);
+		}
+
+		$request_nonce  = null;
+		$authorizations = array_map( 'trim', explode( ',', wp_unslash( $_SERVER['HTTP_AUTHORIZATION'] ) ) );
+		$pattern        = sprintf( '/^%s\s+(?P<nonce>\w+)$/', self::VALIDATE_AUTHORIZATION_TYPE );
+		foreach ( $authorizations as $authorization ) {
+			if ( preg_match( $pattern, $authorization, $match ) ) {
+				$request_nonce = $match['nonce'];
+				break;
+			}
+		}
+
+		if ( empty( $request_nonce ) ) {
+			return new WP_Error(
+				'http_request_failed',
+				esc_html__( 'Missing authorization header.', 'amp' )
+			);
+		}
+
+		if ( $validate_nonce !== $request_nonce ) {
 			return new WP_Error(
 				'http_request_failed',
-				__( 'Nonce authentication failed.', 'amp' )
+				sprintf(
+					/* translators: %s is the WP-CLI command to generate the nonce */
+					esc_html__( 'Nonce authentication failed. The required nonce is returned by %s.', 'amp' ),
+					'<code>wp amp validation generate-nonce</code>'
+				)
 			);
 		}
 
@@ -1817,7 +1857,7 @@ class AMP_Validation_Manager {
 	public static function validate_url( $url ) {
 
 		$added_query_vars = [
-			self::VALIDATE_QUERY_VAR   => self::get_amp_validate_nonce(),
+			self::VALIDATE_QUERY_VAR   => '',
 			self::CACHE_BUST_QUERY_VAR => wp_rand(),
 		];
 		$validation_url   = add_query_arg( $added_query_vars, $url );
@@ -1836,6 +1876,7 @@ class AMP_Validation_Manager {
 					'redirection' => 0, // Because we're in a loop for redirection.
 					'headers'     => [
 						'Cache-Control' => 'no-cache',
+						'Authorization' => sprintf( '%s %s', self::VALIDATE_AUTHORIZATION_TYPE, self::get_amp_validate_nonce() ),
 					],
 				]
 			);
```